#169257396  Signin and Signup validations

Mastel22 opened the pull request:
#### What does this PR do?
Creates validation schemas for signin and signup
#### Description of Task to be completed?
Added:

- [X] validation schemas
- [X] validation schema middleware
#### How should this be manually tested?

- Git clone this repo
- Run `npm i`
- Run `npm run watch`
- Use Postman 

#### Any background context you want to provide?
-N/A
#### What are the relevant pivotal tracker stories?
[#169257396](https://www.pivotaltracker.com/story/show/169257396)
#### Screenshots

<img width="1144" alt="image" src="https://user-images.githubusercontent.com/52410156/69052233-38166380-0a0f-11ea-8e47-636ef4b737dd.png">

![image (1)](https://user-images.githubusercontent.com/52410156/68764754-f73fd880-0623-11ea-96ed-33a58597424f.png)

#### Questions:
None

Mastel22 committed:
feature(validation):Signin and Signup validations

- added the signup schema
- added the signin schema

Mastel22 committed:
feature(validation):Signin and Signup validations

- added the signup schema
- added the signin schema
- added function to test if Email exists

[finishes #169257396]

RaymondMwaura reviewed (commented):
Maybe you can move the validation file to the validation folder as shown on the sample project structure below:
https://github.com/andela/learning-facilitators-bn-backend/tree/master/src

nshaphan committed:
feature(signup): register and generate token
- Create user model and migrations
- Create seeds
- Validate request parameters
- Hash password and store user
- Generate JWT token and return response

[Finishes #169257394]

timolinn commented:
> #### How should this be manually tested?
> * Git clone this repo
> 
Please specify how this should be tested.

timolinn commented on src/middlewares/user.validation.js:
```diff
@@ -0,0 +1,13 @@
+import Responses from '../utils/response';
+
+// eslint-disable-next-line require-jsdoc
+const findByEmail = async (req, res, next) => {
```
Good job here Mastel! I feel we could have here, a name that is more **reflective** of what this function is doing?

timolinn replied:
Also add JSDoc here. There's probably an extension in your IDE that could help with that.

timolinn commented on src/middlewares/validation.js:
```diff
@@ -0,0 +1,30 @@
+      next();
+    });
+  }
+  res.status(405).json({
```
You can use the Response handler for this. Check #18

Mastel22 replied:
Done now

timolinn commented on src/middlewares/user.validation.js:
```diff
@@ -0,0 +1,19 @@
+import Responses from '../utils/response';
+
+/**
+ * Finds if the email passed in the body already exists
```
Change `Finds` to `Checks`.

Mastel22 replied:
Done.

b0nbon1 reviewed (dismissed):
Great work

nshaphan reviewed (dismissed):
LGTM

timolinn reviewed (commented):
Well done Mastel, it all looks good to me except that we cannot really test this because the CreateUser endpoint is still non-functional.

Mastel22 commented:
> Well done Mastel, it all looks good to me except that we cannot really test this because the CreateUser endpoint is still non-functional.

RaymondMwaura reviewed (changes requested):
This works as expected. However, I think you can have it return all the errors in an array form so that the user knows all the errors that they have to fix in one response.

Mastel22 committed:
ft(validations):Restructuring validation files and message

- moved the validation files from the middleware into the validation folder
- mapped the errors from the schemas so that they can be displayed all in an array

Mastel22 commented:
> This works as expected. However, I think you can have it return all the errors in an array form so that the user knows all the errors that they have to fix in one response.

Sorry it took a while to respond because of the restrictions, but it's resolved now

b0nbon1 reviewed (dismissed):
Tested and I have confirmed its working. Great job.
I think it is ready to be merged

timolinn commented:
Good job Mastel, please fix conflicts.

RaymondMwaura reviewed (commented):
This is well done. Fix conflicts and we'll be good to go.

b0nbon1 commented:
Great work

nshaphan reviewed (approved):
Looks good!

RaymondMwaura reviewed (approved):
The conflicts are resolved. Good work.
